rust/src/array_data.rs

```diff
@@ -0,0 +1,264 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use bitmap::Bitmap;
+use buffer::Buffer;
+use datatypes::DataType;
+use util::bit_util::count_set_bits;
+
+#[derive(PartialEq, Debug)]
+pub struct ArrayData {
+    /// The data type for this array data
+    data_type: DataType,
+
+    /// The number of elements in this array data
+    length: i64,
+
+    /// The number of null elements in this array data
+    null_count: i64,
+
+    /// The offset into this array data.
+    offset: i64,
+
+    /// The buffers for this array data. Note that depending on the array types, this
+    /// could hold different types of buffers (e.g., value buffer, value offset buffer)
+    /// at different positions.
+    buffers: Vec<Buffer>,
+
+    /// The child(ren) of this array. Only non-empty for `ListArray` and `StructArray`.
+    child_data: Vec<ArrayDataRef>,
+
+    /// The null bit map. A `None` value for this indicates all values are non-null in
+    /// this array.
+    null_bitmap: Option<Bitmap>,
+}
+
+pub type ArrayDataRef = Arc<ArrayData>;
+pub const UNKNOWN_NULL_COUNT: i64 = -1;
+
+impl ArrayData {
+    pub fn new(
+        data_type: DataType,
+        length: i64,
+        mut null_count: i64,
+        null_bit_buffer: Option<Buffer>,
+        offset: i64,
+        buffers: Vec<Buffer>,
+        child_data: Vec<ArrayDataRef>,
+    ) -> Self {
+        if null_count < 0 {
+            null_count = if let Some(ref buf) = null_bit_buffer {
+                count_set_bits(buf.data())
+            } else {
+                0
+            }
+        }
+        let null_bitmap = null_bit_buffer.map(Bitmap::from);
+        Self {
+            data_type,
+            length,
+            null_count,
+            offset,
+            buffers,
+            child_data,
+            null_bitmap,
+        }
+    }
+
+    /// Returns a builder to construct a `ArrayData` instance.
+    pub fn builder(data_type: DataType) -> ArrayDataBuilder {
+        ArrayDataBuilder::new(data_type)
+    }
+
+    /// Returns a copy of the data type for this array data.
+    pub fn data_type(&self) -> DataType {
+        self.data_type.clone()
```

Comment:
I'm curious, why is this a copy? Data types are mutable?

Reply:
Data types are not mutable. The other option is to return a reference here. However, if you need to store this reference anywhere in a struct, you'd have to add a lifetime parameter in that struct, which will get populated to any struct that use this struct, and so on. Personally I'd like to avoid this since it makes the design much more restrictive.

I think it's a good idea to add another version of this method that returns a reference though, in case anyone only needs to use the result in the local stack. Will do that.

Reply:
(IMHO!) In Rust, the user should decide if the clone is required or not. In many cases, you could read out information of the data type w/o cloning the entire type. It would follow the "zero-cost abstraction"-principle. And for `buffers` and `child_data` you do the same, so it would be consistent. Also, `data_type_ref` is not really "cheap/efficient by default".

Reply:
Got you point. I think we can remove the `data_type` and rename `data_type_ref` to `data_type`. This keep it consistent with others like `buffer` and `child_data`. What do you think?